### PR TITLE
Whisper: Fix decoder inputs for static pipeline

### PR DESCRIPTION
**Pull request by eshiryae**
### Tickets:
* [CVS-159064](https://jira.devtools.intel.com/browse/CVS-159064)

### CODE REVIEW

File: src/cpp/src/whisper_pipeline_static.hpp [hunk 1]
```diff
@@ -25,6 +25,7 @@ class WhisperPipeline::StaticWhisperPipeline : public WhisperPipeline::WhisperPi
 
 private:
     WhisperInitializedModels m_models;
+    std::shared_ptr<ov::Model> m_decoder_model;
```
Comment (ilya-lavrenov) [inline src/cpp/src/whisper_pipeline_static.hpp:28]: why do we need to store model? once model is compiled, we need to release ov::Model to free memory consumed by its weights

Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.hpp:28]: We can't compile this model until `generate()` called.

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 2]
```diff
@@ -210,13 +210,13 @@ void zero_past_key_values(ov::InferRequest& request) {
...
-void prepare_decoder_with_past(ov::InferRequest& decoder_with_past, ov::InferRequest& decoder) {
+void prepare_decoder_with_past(ov::InferRequest& decoder_with_past, ov::InferRequest& decoder, const size_t init_ids_size) {
     // NB: Prepare attetion mask to be in a format [0, 0, 0, 1, 1, 1, 1, ..., 0, 1]
     // Mask should be inverted for decoder_with_past 
     auto attention_mask = decoder_with_past.get_tensor("attention_mask");
     auto* attention_mask_ptr = attention_mask.data<ov::float16>();
-    std::fill(attention_mask_ptr, attention_mask_ptr + 3u, 0);
-    std::fill(attention_mask_ptr + 3u, attention_mask_ptr + attention_mask.get_size() - 2, 1);
+    std::fill(attention_mask_ptr, attention_mask_ptr + init_ids_size, 0);
+    std::fill(attention_mask_ptr + init_ids_size, attention_mask_ptr + attention_mask.get_size() - 2, 1);
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:219]: Why `-2`?

Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.cpp:219]: Btw, since we reshape model anyway, we don't need `attention_mask` at all, probably we may not apply transformation that expose this

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 3]
```diff
@@ -489,7 +489,7 @@ void preprocess_decoder(std::shared_ptr<ov::Model> model) {
             preprocessor.input("attention_mask").preprocess().convert_element_type();
         } else if (tensor.get_any_name().find("encoder_hidden_states") != std::string::npos) {
             preprocessor.input("encoder_hidden_states").tensor().set_element_type(ov::element::Type_t::f16);
-            preprocessor.input("encoder_hidden_states").preprocess().convert_element_type(ov::element::Type_t::f32); // ()
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:492]: Why it's removed?

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 4]
```diff
@@ -654,7 +657,13 @@ WhisperDecodedResults WhisperPipeline::StaticWhisperPipeline::generate(
 
         // prepare init_ids just once for whole input
         if (init_ids.empty()) {
+            OPENVINO_ASSERT(m_models.decoder.get_tensor("input_ids").get_shape().back() == 1);
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:660]: What does  this check do?

**Review by TolyaTalamanov — COMMENTED** [review]
Well, I'd assume we have something like this:

```
class DecoderCache {
public:
    ov::CompiledModel get_model(uint8_t input_id_size) {
        // Get from hash table, otherwise compile and store...
    }
private:
    // [input_ids_size -> CompiledModel]
    std::unordered_map<uint8_t, ov::CompiledModel> m_cache;
    std::shared_ptr<ov::Model> decoder_model; // <- this is dynamic w/o transformation applied    
}


// Whenever we need a model:
auto decoder = m_decoder_cache.get(input_ids_size);
```

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 3 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp:492]: by mistake

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 2 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp:219]: This function is for decoder_with_past model.
It should have attention mask should be in format [0, 0, 0, 1, 1, 1, ... 1, 0, 1]

You mean we don't need attention mask for decoder model or for both model? How will it work then for decoder with past model?

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 5]
```diff
@@ -136,9 +136,9 @@ void set_decoder_input_ids_attention_mask(ov::InferRequest& decoder,
     // attention_mask [1, 1, 1, 0]
     auto input_ids_data = input_ids_tensor.data<int32_t>();
     std::copy(init_ids.begin(), init_ids.end(), input_ids_data);
-    std::fill(input_ids_data + init_ids.size(),
-              input_ids_data + input_ids_tensor.get_size(),
-              static_cast<int32_t>(pad_token));
+    // std::fill(input_ids_data + init_ids.size(),
```
[the remark this answers is not in the document]
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp:139]: Sure, will remove

File: src/cpp/src/whisper_pipeline_static.hpp (hunk 1 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.hpp:28]: It's a temporal solution as wee need to reshape and recompile decoder model for specific number of input tokens, but we know it only on generation stage.

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.cpp:657]: I assume it can't be incorrect, as we explicitly reshape model for `input_ids` size.

Besides, I'd rather check: `get_tensor("input_ids").get_size()`

**Comment by eshiryae** [comment]
1. `check_decoder_model_compatibility()` checks only if decoder_model has attention_mask as an input, shape will not affects this check. Btw i suppose we don't need this check and we can remove it.

Agree with other points, it's a great idea to store already compiled models in cache, will do that.

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 2 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp:216]: There's already such comment here: 
https://github.com/openvinotoolkit/openvino.genai/blob/5ab58ca70dd2774595ad82768074c7a497aa9377/src/cpp/src/whisper_pipeline_static.cpp#L214-L215

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 6]
```diff
@@ -541,6 +514,17 @@ std::shared_ptr<ov::Model> redirect_new_kv_to_output(const std::shared_ptr<ov::M
 namespace ov {
 namespace genai {
 
+ov::CompiledModel DecoderCache::get_model(uint8_t input_ids_size) {
+    if (m_cache.find(input_ids_size) == m_cache.cend()) {
+        reshape_input_ids(m_decoder_model, input_ids_size);
+
+        ov::Core core = utils::singleton_core();
+        m_cache.insert({input_ids_size, core.compile_model(m_decoder_model, "NPU")});
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:522]: Nitpick: `m_cache.emplace(input_ids_size, core.compile_model(m_decoder_model, "NPU"))l`

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 7]
```diff
@@ -550,20 +534,14 @@ WhisperPipeline::StaticWhisperPipeline::StaticWhisperPipeline(const std::filesys
...
-        OPENVINO_THROW("StaticWhisperPipeline expects decoder model has \"attention_mask\" input!");
-    }
-
     size_t max_sequence_length = 448;
 
     reshape_to_static_encoder(encoder_model, m_feature_extractor.feature_size);
 
     auto last_hidden_state_shape = get_encoder_hidden_state_shape(encoder_model);
-    reshape_to_static(decoder_model, 4, 4, last_hidden_state_shape);
+    reshape_to_static(decoder_model, 1, 1, last_hidden_state_shape); // for detect_language()
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:544]: I believe you no longer need to reshape model anywhere, it's responsibility of `DecoderCache`

File: src/cpp/src/whisper_pipeline_static.hpp [hunk 8]
```diff
@@ -15,6 +15,17 @@
 namespace ov {
 namespace genai {
 
+class DecoderCache {
+public:
+    DecoderCache() {}
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.hpp:20]: `DecoderCache() = default`

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 7 above)
Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.cpp:544]: If you need it to detect language, call `DecoderCache` when you really needed

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 9]
```diff
@@ -467,6 +424,22 @@ void reshape_to_static_encoder(std::shared_ptr<ov::Model> model, const size_t fe
...
 
+void reshape_input_ids(std::shared_ptr<ov::Model> model, const uint32_t input_size) {
+    std::map<std::string, ov::PartialShape> new_shapes;
+    for (auto input : model->inputs()) {
+        const auto& input_name = input.get_any_name();
+        ov::PartialShape new_shape;
+        if (input_name.find("input_ids") != std::string::npos) {
+            new_shape = ov::PartialShape({1, input_size});
+        } else {
+            new_shape = input.get_partial_shape();
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:435]: Can you just pass one layer to `reshape` function? It's weird you need to duplicate layers that are not supposed to be changed

**Review by TolyaTalamanov — APPROVED** [review]

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 6 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp]: fixed

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 7 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp]: done

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 9 above)
Comment (eshiryae) [reply src/cpp/src/whisper_pipeline_static.cpp]: fixed

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 10]
```diff
@@ -654,7 +626,11 @@ WhisperDecodedResults WhisperPipeline::StaticWhisperPipeline::generate(
 
         // prepare init_ids just once for whole input
         if (init_ids.empty()) {
+            m_models.decoder = m_decoder_cache.get_model(1).create_infer_request(); // for detect_language()
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:629]: Here the model compiled upfront again. What if it's not even needed and language already known?

I believe you need it somewhere here: https://github.com/openvinotoolkit/openvino.genai/blob/614e6d9b7802d89b00c480e20b27c0f5b7913b04/src/cpp/src/whisper_pipeline_static.cpp#L269

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 11]
```diff
@@ -541,6 +502,23 @@ std::shared_ptr<ov::Model> redirect_new_kv_to_output(const std::shared_ptr<ov::M
 namespace ov {
 namespace genai {
 
+ov::CompiledModel DecoderCache::get_model(uint8_t input_ids_size) {
+    if (m_cache.find(input_ids_size) == m_cache.cend()) {
+        if (m_decoder_model->is_dynamic()) { // model is dynamic, reshaping it to static
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:507]: Weird, I don't expect we need this check...

Let's discuss locally

File: src/cpp/src/whisper_pipeline_static.hpp [hunk 12]
```diff
@@ -15,6 +15,20 @@
 namespace ov {
 namespace genai {
 
+class DecoderCache {
+public:
+    DecoderCache() = default;
+    DecoderCache(std::shared_ptr<ov::Model> model, ov::PartialShape shape)
+     : m_decoder_model(model)
+     , m_lhs_shape(shape) {}
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.hpp:23]: Well, `decoder` model has two input layers:
 1. `encoder_hidden_states` - does it change? I believe it depends on the `encoder` model and once we know it it's no longer change?
 2. `decoder_input_ids` - This is what we track in hash map. Changes depends on the case
 
 Can we set `encoder_hidden_states` in the `ctor`? If so, I believe we don't need to change if model `dynamic` or not in the `get()` method

Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.hpp:23]: I believe the size of `encoder_hiddne_states` depends only on `feature_size` and it's set in `StaticWhisperPipeline` and no longer changes since then: https://github.com/openvinotoolkit/openvino.genai/blob/614e6d9b7802d89b00c480e20b27c0f5b7913b04/src/cpp/src/whisper_pipeline_static.cpp#L535. Can you confirm this?

If so, I believe we definitely can set shape for `encoder_hidden_states` once.

```
// Also, it should probably be `ov::Shape` rather than ov::PartialShape as we don't have dynamism (input is fixed)
DecoderCache(std::shared_ptr<ov::Model> model, ov::PartialShape encoder_hidden_shape) {
    model->reshape({{"encoder_hidden_states", encoder_hidden_shape}});
}
DecoderCache::get(ov::CompiledModel DecoderCache::get_model(uint8_t input_ids_size) {
    if (m_cache.counts(input_ids_size) == 0) {
        m_model->reshape({{"decoder_input_ids", ov::Shape({1, input_ids_size})}});
        m_cache.emplace(input_ids_size, core.compile_model(m_model));
    }
    return m_cache.at(input_ids_size);
}
```
